src/cargo/ops/resolve.rs

```diff
@@ -198,10 +198,32 @@ pub fn resolve_ws_with_opts<'cfg>(
         &member_ids,
         has_dev_units,
         requested_targets,
-        target_data,
+        &mut *target_data,
         force_all_targets,
     )?;
 
+    // After we download all the packages we need, we need to change target_data such that any
+    // additionally enabled targets are registered, since they could have per-pkg-target or
+    // artifact dependencies.
+    for pkg in pkg_set.packages() {
+        fn artifact_targets(package: &Package) -> impl Iterator<Item = CompileKind> + '_ {
+            package
+                .manifest()
+                .dependencies()
+                .iter()
+                .filter_map(|d| d.artifact()?.target()?.to_compile_kind())
+        }
+        let kinds = pkg
+            .manifest()
+            .default_kind()
+            .into_iter()
+            .chain(pkg.manifest().forced_kind())
+            .chain(artifact_targets(pkg));
```

Comment:
I'm curious, what situations would need to merge `artifact_targets` here? My initial thought is that they should have all already been merged during `download_accessible`.